pkg/operator/resource/resourceapply/apps.go

```diff
@@ -15,13 +18,42 @@ import (
 	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
 )
 
-// ApplyDeployment merges objectmeta and requires matching generation. It returns the final Object, whether any change as made, and an error
-func ApplyDeployment(client appsclientv1.DeploymentsGetter, recorder events.Recorder, required *appsv1.Deployment, expectedGeneration int64,
+const specHashAnnotation = "operator.openshift.io/spec-hash"
+
+// ApplyDeployment merges ObjectMeta of the provided deployment with an existing one
+// if it exists and updates the API if the deployment spec and metadata differ. To
+// ensure an update in response to state external to the deployment spec, the caller
+// should set an annotation representing that external state.
+func ApplyDeployment(client appsclientv1.DeploymentsGetter, recorder events.Recorder,
+	required *appsv1.Deployment) (*appsv1.Deployment, bool, error) {
+
+	// Record the hash of the spec in an annotation to prompt rollout if it changes.
+	jsonBytes, err := json.Marshal(required.Spec)
+	if err != nil {
+		return nil, false, err
+	}
+	specHash := fmt.Sprintf("%x", sha256.Sum256(jsonBytes))
+	if required.Annotations == nil {
+		required.Annotations = map[string]string{}
+	}
+	required.Annotations[specHashAnnotation] = specHash
+
+	return ApplyDeploymentWithForce(client, recorder, required, 0, false)
```

Comment:
I think you need an expected generation in order to know if another actor mutated the spec.  The hash of the input spec lets you know if the operator decided a new value was needed (--loglevel for instance) and the generation lets you know if an external actor changed the deployment spec. 

Because defaults can change on the server (new field appearing in a new level of kube for instance), you cannot reliably read and compare a previous hash.

Reply:
Done